### PR TITLE
Review: CDAP 19962 Connection List Page

### PR DESCRIPTION
**Pull request by adityadivami**
# TITLE
Feature - Connection List Page

## Description
Summary of changes
This PR contains the changes related to whole Connection List Page

## PR Type
- [ ] Bug Fix
- [x] Feature
- [ ] Build Fix
- [ ] Testing
- [ ] General Improvement
- [ ] Cherry Pick

## Links
Jira: [Jira issue #](fill-in.org)
https://cdap.atlassian.net/browse/CDAP-19962

## Files to be reviewed as part of this PR
1. app/cdap/components/ConnectionList/*
2. app/cdap/components/DataPrep/store/index.ts
3. app/cdap/components/EntityTopPanel/EntityCloseBtn.tsx
4. app/cdap/text/text-en.yaml
5. src/e2e-test/*

## Test Plan
1. FE_03_310 - Verify whether the User is redirected to Data Source Page or not. - Pass
2. FE_03_311 - Verify whether the "Add Connection" text & "+" Icon are displayed at the right end of the header as per design or not. - Pass
3. FE_03_312 - Verify by mouse hovering on "+" icon - Pass
4. FE_03_313 - Verify by clicking on "+" icon - Pass
5. FE_03_314 - Verify the "Add Connection" text by mouse hovering. - Pass
6. FE_03_315 - Verify whether the "Add Connection" text link is clickable or not. - Pass
7. FE_03_316 - Verify after clicking on the "Add Connection" link. - Pass
8. FE_03_317 - Verify after creating the new connection, it redirects to the data source page or not. - Pass
9. FE_03_318 - Verify after clicking on the close icon, the link redirects to the data source page or not. - Pass
10. FE_03_319 - Verify whether the "Data Connections" text is displayed as the title of the first column or not. - Pass
11. FE_03_320 - Verify by clicking on the "Data Connections" text. - Pass
12. FE_03_321 - Verify by mouse hovering on the "Data Connections" text of the first column. - Pass
13. FE_03_322 - Verify the connection types from the data connection column are clickable or not  - Pass
14. FE_03_323 - Verify the redirection after clicking on the connection types from the first connection column. - Pass
15. FE_03_324 - Verify the selected list from the column is displayed in blue color or not. - Pass
16. FE_03_325 - Verify whether the horizontal columns are automatically updating with every selection of the list or not. - Pass
17. FE_03_326 - Verify the flow of the table column. - Pass
18. FE_03_327 - Verify whether the count is displayed beside the connector type text or not. - Pass
19. FE_03_328 - Verify the Alignments & paddings between one data connection type to another type are displayed as per design or not. - Pass
20. FE_03_329 - Verify whether the bucket(File/Folder) column is Displayed as per design or not. - Pass
21. FE_03_330 - Verify the paddings of all columns are as per Figma or not. - Pass
22. FE_03_331 - Verify the long names are displayed with ellipses(tooltip) in every column - Pass
23. FE_03_332 - Verify by mouse hovering on Connection types from the data connection column. - Pass
24. FE_03_333 - Verify whether the vertical scroll bar is displayed or not for the individual columns (if there are more files) - Pass
25. FE_03_334 - Verify whether the horizontal Scroll bar is automatically updating with every new column opening or not. - Pass
26. FE_03_335 - Verify whether the horizontal columns have a search icon beside the updated columns as per Figma or not. - Pass
27. FE_03_336 - Verify the Search icon beside the updated column by mouse hovering. - Pass
28. FE_03_337 - Verify the Search icon beside the updated columns by Clicking. - Pass
29. FE_03_338 - Verify the Search icon beside the updated columns by Editing. - Pass
30. FE_03_339 - Verify whether the expected search result is displaying or not from the column. - Pass

## Screenshots
Screenshot 2022-11-21 at 8 36 05 PM
Screenshot 2022-11-21 at 8 36 48 PM
Screenshot 2022-11-21 at 8 37 16 PM

### CODE REVIEW

File: app/cdap/components/ConnectionList/Components/ConnectionTabs/Components/RenderLabel/index.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+export interface IRenderLabelProps {
+  columnIndex: number;
+  connectorType: IConnectionTabType;
+  connectionIdProp: string;
+  toggleLoader: (value: boolean, isError?: boolean) => void;
+  setIsErrorOnNoWorkSpace: Dispatch<SetStateAction<boolean>>;
+  dataTestID: number;
+}
+
+export interface IConnectionTabType {
```
Comment (njbriggs) [inline app/cdap/components/ConnectionList/Components/ConnectionTabs/Components/RenderLabel/index.tsx:31]: This and the following types seem broadly applicable - why define them here?

Comment (njbriggs) [inline app/cdap/components/ConnectionList/Components/ConnectionTabs/Components/RenderLabel/index.tsx]: This and the following types seem broadly applicable - why define them here?

File: app/cdap/components/ConnectionList/Components/ConnectionTabs/__test__/ConnectionTabs.test.tsx [hunk 2]
```diff
@@ -14,25 +14,125 @@
...
+
   it('Should render Connections Tabs Parent Component', () => {
     render(
-      <ConnectionsTabs
+      <ConnectionTabs
         tabsData={mockTabsTestData}
         handleChange={() => null}
         value="apple"
-        index="0"
+        // index={0}
```
Comment (njbriggs) [inline app/cdap/components/ConnectionList/Components/ConnectionTabs/__test__/ConnectionTabs.test.tsx]: Remove this?

File: app/cdap/components/ConnectionList/Components/ConnectionTabs/index.tsx [hunk 3]
```diff
@@ -14,128 +14,157 @@
...
+    }
+  }, [refValue]);
+
   return (
-    <Box data-testid="connections-tabs-parent" className={classes.connectionsTabsParent}>
-      {tabsData.showTabs && (
-        <div className={classes.boxStyles}>
-          <Tabs
+    <Box data-testid="connections-tabs-parent" {...{ ref: refValue }}>
+      {tabsData?.showTabs && (
```
Comment (njbriggs) [inline app/cdap/components/ConnectionList/Components/ConnectionTabs/index.tsx]: Why the `?`? - `tabsData` is not marked as nullable

File: app/cdap/components/ConnectionList/Components/Header/Components/HeaderLabelWrapper/index.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+}: IHeaderCustomTooltipProps) {
+  const tooltipRequired =
+    headersRefs?.current[columnIndex]?.offsetWidth < headersRefs?.current[columnIndex]?.scrollWidth;
+
+  return (
+    <CustomTooltip title={tooltipRequired ? tabsData[columnIndex - 1].selectedTab : ''} arrow>
+      <RenderLabel
+        variant="body2"
+        ref={(element) => {
+          headersRefs.current[columnIndex] = element;
```
Comment (njbriggs) [inline app/cdap/components/ConnectionList/Components/Header/Components/HeaderLabelWrapper/index.tsx]: Please do not mutate `props`.

Why does `headerRefs` need to be a prop anyway? It looks like it's never accessed outside this component.

File: app/cdap/components/ConnectionList/Components/Header/Components/HeaderLabelWrapper/index.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        }}
+      >
+        {filteredData[columnIndex - 1]?.selectedTab}
```
Comment (njbriggs) [inline app/cdap/components/ConnectionList/Components/Header/Components/HeaderLabelWrapper/index.tsx]: Are `tabsData` and `filteredData` the same? If not, will accessing the same index give the same results? If they are, why pass both in?

File: app/cdap/components/ConnectionList/IconStore/CGSIcon.tsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import React from 'react';
+
+export const GCSIcon = () => {
```
Comment (njbriggs) [inline app/cdap/components/ConnectionList/IconStore/CGSIcon.tsx]: Why is this icon here? It should be fetched from the API.

File: app/cdap/components/ConnectionList/index.tsx [hunk 7]
```diff
@@ -46,228 +90,294 @@ const SelectDatasetWrapper = styled(Box)({
...
+   */
+  DataPrepStore.subscribe(() => {
+    const newState = DataPrepStore.getState();
+    setFetchedConnectorsData(newState.dataprep.connectorsWithIcons);
+  });
 
   const getConnectionsTabData = async () => {
-    // Fetching the all available connectors list
-    let connectorTypes = await fetchConnectors();
+    let connectorTypes: ITabData[] = [];
```
Comment (njbriggs) [inline app/cdap/components/ConnectionList/index.tsx]: Why not `let connectorTypes = fetchedConnectorsData`?

File: app/cdap/components/ConnectionList/index.tsx [hunk 8]
```diff
@@ -46,228 +90,294 @@ const SelectDatasetWrapper = styled(Box)({
...
     let allConnectionsTotalLength = 0;
-
-    // Fetching all the connections list inside a connector
+    // Fetching all the connections list inside each connector type
     const categorizedConnections = await getCategorizedConnections();
-    connectorTypes = connectorTypes.filter((conn) => {
-      return [conn.name];
+    const connectorTypeWithConnections = [];
+    categorizedConnections?.forEach((itemEach, key) => {
+      connectorTypeWithConnections.push(key);
```
Comment (njbriggs) [inline app/cdap/components/ConnectionList/index.tsx:160]: Make this a `map`?

File: app/cdap/components/ConnectionList/index.tsx [hunk 9]
```diff
@@ -46,228 +90,294 @@ const SelectDatasetWrapper = styled(Box)({
...
+    });
+
+    // filtering all the connections which got connections inside them
+    connectorTypes = connectorTypes.filter((obj) =>
+      connectorTypeWithConnections.find((item) => item == obj.name)
+    );
+    connectorTypes = [...new Map(connectorTypes.map((item) => [item.name, item])).values()];
+    connectorTypes = connectorTypes.filter((eachConnectionType: ITabData) => {
```
Comment (njbriggs) [inline app/cdap/components/ConnectionList/index.tsx]: This sequence is hard to follow. Can this be done in a single expression? Or if not, use a different variable for each step.

**Review by njbriggs — CHANGES_REQUESTED** [review]

File: app/cdap/components/ConnectionList/Components/ConnectionTabs/__test__/ConnectionTabs.test.tsx (hunk 2 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/Components/ConnectionTabs/__test__/ConnectionTabs.test.tsx]: Done. Removed this commented line.

File: app/cdap/components/ConnectionList/Components/ConnectionTabs/index.tsx (hunk 3 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/Components/ConnectionTabs/index.tsx:135]: Here '?' is not required as the tabsData is getting fetched as a prop from connectionList index.tsx component.
So, we have removed the '?' here and updated the code.

File: app/cdap/components/ConnectionList/Components/Header/Components/HeaderLabelWrapper/index.tsx (hunk 5 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/Components/Header/Components/HeaderLabelWrapper/index.tsx]: No, tabsData and filteredData are not same. But here in this scenario we trying to get the previous level's selected tab value. So, it doesn't matter if we use tabsData or filteredData here. Updated the code & removed passing tabsData as a prop to this component.

File: app/cdap/components/ConnectionList/IconStore/CGSIcon.tsx (hunk 6 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/IconStore/CGSIcon.tsx]: This icon has been used in the test files for mocking data but we have modified the test files now & removed this static icon.

File: app/cdap/components/ConnectionList/index.tsx (hunk 9 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/index.tsx]: Modified the code by providing different variable names for each step.

File: app/cdap/components/ConnectionList/index.tsx (hunk 8 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/index.tsx:160]: Here we have used forEach instead of map method because map method doesn't work on categorizedConnections as it is a map object.

File: app/cdap/components/ConnectionList/index.tsx (hunk 7 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/index.tsx]: we modified the code to reflect the above suggestion.

File: app/cdap/components/ConnectionList/Components/ConnectionTabs/Components/RenderLabel/index.tsx (hunk 1 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/Components/ConnectionTabs/Components/RenderLabel/index.tsx]: All these interfaces are getting used in this component. That was why we defined all the required interfaces for this component in this file itself and imported these individual interfaces wherever we needed as well them from here. 

Now, we moved all these types to types.ts of conectionList component & imported the required type from there.

File: app/cdap/components/ConnectionList/Components/ConnectionTabs/index.tsx (hunk 3 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/Components/ConnectionTabs/index.tsx]: removed '?' here as it is not required here as the tabsData is getting fetched as a prop from its parent component. So, it won't be nullable at any point.

File: app/cdap/components/ConnectionList/Components/Header/Components/HeaderLabelWrapper/index.tsx (hunk 4 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/Components/Header/Components/HeaderLabelWrapper/index.tsx]: The code has been updated. Moved this code to its parent component and deleted this file.

File: app/cdap/components/ConnectionList/IconStore/InfoGraph.tsx [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import React from 'react';
+
+export const InfoGraph = () => {
```
Comment (seanfeld) [inline app/cdap/components/ConnectionList/IconStore/InfoGraph.tsx]: What is this svg? is it that cog thing on the first page? I don't think we want an image like that it doesn't fit with anything else in gcp.

Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/IconStore/InfoGraph.tsx]: No, @seanfeld. This SVG appears on the second page(connection list page) at right bottom side. (Please refer to the first screenshot which is attached in the description)

File: app/cdap/components/ConnectionList/IconStore/WrangleIcon.tsx [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import React from 'react';
+
+export const WrangleIcon = () => (
```
Comment (seanfeld) [inline app/cdap/components/ConnectionList/IconStore/WrangleIcon.tsx]: don't we already have this icon used somewhere?

Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/IconStore/WrangleIcon.tsx]: No @seanfeld. We have not used this wrangle icon anywhere else. It is used in the connection list page only to indicate a dataset that can be loaded to the grid.

File: app/cdap/components/ConnectionList/index.tsx [hunk 12]
```diff
@@ -46,228 +94,295 @@ const SelectDatasetWrapper = styled(Box)({
...
+  const headersRefs = useRef([]);
+  const [loading, setLoading] = useState<boolean>(true);
   const [isErrorOnNoWorkspace, setIsErrorOnNoWorkSpace] = useState<boolean>(false);
-
-  const toggleLoader = (value: boolean, isError?: boolean) => {
-    setLoading(value);
-  };
-  let connectionId;
-  const [dataForTabs, setDataForTabs] = useState([
+  const [tabsLength, setTabsLength] = useState<number>(0);
```
Comment (seanfeld) [inline app/cdap/components/ConnectionList/index.tsx]: if you have more than 5 useStates you should convert it to a useReducer

File: app/cdap/components/ConnectionList/index.tsx [hunk 13]
```diff
@@ -46,228 +94,295 @@ const SelectDatasetWrapper = styled(Box)({
...
+  let connectionId;
+
+  useEffect(() => {
+    getWidgetData();
+  }, []);
+
+  /**
+   * Here we are fetching connectors with icons from store
+   */
+  DataPrepStore.subscribe(() => {
```
Comment (seanfeld) [inline app/cdap/components/ConnectionList/index.tsx:144]: this should be in some kind of useEffect or inside a custom hook or using redux

File: app/cdap/components/ConnectionList/types.ts [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+  dataTestID: number;
+}
+
+export interface IConnectionTabType {
+  connectionId?: string;
+  connectionType?: string;
+  createdTimeMillis?: number;
+  description?: string;
+  isDefault?: boolean;
+  name: string;
```
Comment (seanfeld) [inline app/cdap/components/ConnectionList/types.ts:112]: required types should come before optional ones.

File: app/cdap/components/ConnectionList/utils.ts [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+import { IFilteredData, ITabData, ITabsDataResponse } from 'components/ConnectionList/types';
+import {
+  fetchAllConnectorPluginProperties,
+  fetchConnectors,
+  getMapOfConnectorToPluginProperties,
+  getSelectedConnectorDisplayName,
+} from 'components/Connections/Create/reducer';
+
+/**
+ *
```
Comment (seanfeld) [inline app/cdap/components/ConnectionList/utils.ts:26]: Please add what this function does

File: app/cdap/components/ConnectionList/utils.ts [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+  }
+  tempData[index + 1][`showTabs`] = true;
+  tempData[index + 1][`selectedTab`] = null;
+  tempData[index + 1][`toggleSearch`] = false;
+  tempData[index][`toggleSearch`] = false;
+  return tempData.slice(0, index + 2);
+};
+
+/**
+ *
```
Comment (seanfeld) [inline app/cdap/components/ConnectionList/utils.ts:58]: add a brief summary of what the function does

File: app/cdap/components/ConnectionList/utils.ts [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+      eachTempData.toggleSearch = true;
+    } else {
+      eachTempData.toggleSearch = false;
+    }
+  });
+  return tempData;
+};
+
+/**
+ *
```
Comment (seanfeld) [inline app/cdap/components/ConnectionList/utils.ts:82]: again here

**Review by seanfeld — CHANGES_REQUESTED** [review]
Left comments

File: app/cdap/components/ConnectionList/types.ts (hunk 14 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/types.ts:112]: moved the non-optional types to reflect the above suggestion

File: app/cdap/components/ConnectionList/utils.ts (hunk 15 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/utils.ts:26]: The summary has been added to explain what this function does.

File: app/cdap/components/ConnectionList/utils.ts (hunk 16 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/utils.ts:58]: The summary has been added to explain what this function does.

File: app/cdap/components/ConnectionList/utils.ts (hunk 17 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/utils.ts:82]: The summary has been added to explain what this function does.

File: app/cdap/components/ConnectionList/index.tsx (hunk 13 above)
Comment (sivaRajanaDivami) [reply app/cdap/components/ConnectionList/index.tsx:144]: moved this redux store subscribe code block to useEffect.
